File: backend/app/schemas/video.py
"""Pydantic schemas for video generation endpoints."""

import uuid
from datetime import datetime
from enum import Enum

from pydantic import BaseModel, Field


class VideoStatus(str, Enum):
    """Valid states in the generation state machine."""
    QUEUED = "queued"
    PROCESSING = "processing"
    COMPLETED = "completed"
    FAILED = "failed"
    TIMEOUT = "timeout"
    CANCELLED = "cancelled"


class VideoStyle(str, Enum):
    """Supported style presets."""
    CINEMATIC = "cinematic"
    REALISTIC = "realistic"
    ANIME = "anime"
    THREE_D = "3d"
    PRODUCT = "product"


class AspectRatio(str, Enum):
    """Supported aspect ratios."""
    LANDSCAPE = "16:9"
    PORTRAIT = "9:16"
    SQUARE = "1:1"


class Duration(int, Enum):
    """Supported video durations in seconds."""
    SHORT = 5
    LONG = 10


class VideoGenerateRequest(BaseModel):
    """Request body for POST /api/v1/videos/generate."""
    prompt: str = Field(
        ...,
        min_length=1,
        max_length=2000,
        description="The text prompt describing the desired video",
    )
    enhanced_prompt: str | None = Field(
        default=None,
        description="Optional pre-enhanced prompt (from the enhance endpoint)",
    )
    negative_prompt: str | None = Field(
        default=None,
        max_length=500,
        description="Things to avoid in the generated video",
    )
    style: VideoStyle = Field(
        default=VideoStyle.CINEMATIC,
        description="Visual style preset",
    )
    aspect_ratio: AspectRatio = Field(
        default=AspectRatio.LANDSCAPE,
        description="Video aspect ratio",
    )
    duration: Duration = Field(
        default=Duration.SHORT,
        description="Video duration in seconds",
    )


class VideoResponse(BaseModel):
    """Response schema for a video generation record."""
    id: uuid.UUID
    original_prompt: str
    enhanced_prompt: str | None = None
    negative_prompt: str | None = None
    style: str
    aspect_ratio: str
    duration: int
    provider: str
    model: str
    provider_job_id: str | None = None
    status: str
    video_url: str | None = None
    thumbnail_url: str | None = None
    error_message: str | None = None
    created_at: datetime
    updated_at: datetime
    completed_at: datetime | None = None

    model_config = {"from_attributes": True}


class VideoDeleteResponse(BaseModel):
    """Response for DELETE /api/v1/videos/{id}."""
    deleted: bool = True

File: frontend/src/lib/types/video.ts
/** Video generation types aligned with backend schemas. */

export type VideoStatus = 'queued' | 'processing' | 'completed' | 'failed' | 'timeout' | 'cancelled';
export type VideoStyle = 'cinematic' | 'realistic' | 'anime' | '3d' | 'product';
export type AspectRatio = '16:9' | '9:16' | '1:1';
export type Duration = 5 | 10;

export interface VideoGeneration {
  id: string;
  originalPrompt: string;
  enhancedPrompt?: string | null;
  negativePrompt?: string | null;
  style: VideoStyle;
  aspectRatio: AspectRatio;
  duration: Duration;
  provider: string;
  model: string;
  providerJobId?: string | null;
  status: VideoStatus;
  videoUrl?: string | null;
  thumbnailUrl?: string | null;
  errorMessage?: string | null;
  createdAt: string;
  updatedAt: string;
  completedAt?: string | null;
}

export interface VideoGenerateRequest {
  prompt: string;
  enhanced_prompt?: string | null;
  negative_prompt?: string | null;
  style: VideoStyle;
  aspect_ratio: AspectRatio;
  duration: Duration;
}

/** Maps for UI display labels */
export const STYLE_LABELS: Record<VideoStyle, string> = {
  cinematic: 'Cinematic',
  realistic: 'Realistic',
  anime: 'Anime',
  '3d': '3D',
  product: 'Product Commercial',
};

export const ASPECT_RATIO_LABELS: Record<AspectRatio, string> = {
  '16:9': '16:9 Landscape',
  '9:16': '9:16 Portrait',
  '1:1': '1:1 Square',
};

export const DURATION_LABELS: Record<Duration, string> = {
  5: '5 seconds',
  10: '10 seconds',
};

export const TERMINAL_STATUSES: VideoStatus[] = ['completed', 'failed', 'timeout', 'cancelled'];
